src/engine/worker.cpp

```diff
@@ -18,8 +19,59 @@
 #include "sampling/logits_processor.h"
 #include "sampling/sampler.h"
 
+DEFINE_bool(enable_cudagraph, false, "Enable CUDAGraph to optimize model execution.");
+
 namespace llm {
 
+class CudaGraphRunner {
+ public:
+  explicit CudaGraphRunner(CausalLM* model) : model_(model) {}
+  ~CudaGraphRunner() = default;
+
+  // ids, positions, kv_cache, input_meta, memorypool
+  void capture(torch::Tensor flatten_tokens, torch::Tensor flatten_positions,
+      const InputParameters& params, std::vector<KVCache>& kv_cache) {
+    // run model once to avoid captured graph does not include initial benchmarking
+    model_->forward(flatten_tokens,
+                    flatten_positions,
+                    kv_cache,
+                    params);
+    torch::cuda::synchronize();
+
+    // create cudagraph and capture
+    graph_ = std::make_unique<at::cuda::CUDAGraph>();
+    graph_->capture_begin();
+    auto logits = model_->forward(flatten_tokens,
+                                  flatten_positions,
+                                  kv_cache,
+                                  params);
+    graph_->capture_end();
+    torch::cuda::synchronize();
+
+    // save buffers
+    input_buffer_.emplace("flatten_tokens", flatten_tokens);
+    input_buffer_.emplace("flatten_positions", flatten_positions);
+    // input_buffer_.emplace("kv_cache", kv_cache);
+    // TODO: slot_mapping, context_lens, block_tables
+    output_buffer_.emplace("logits", logits);
+  }
+
+  // ids, positions, kv_cache, input_meta
+  torch::Tensor forward(torch::Tensor flatten_tokens, torch::Tensor flatten_positions,
+      const InputParameters& params, std::vector<KVCache>& kv_cache) {
+    input_buffer_["flatten_tokens"] = flatten_tokens,
```

Comment:
i think you need to copy the value from input faltten_tokens to input_buffer, otherwise, replay would still use the old tensor instead of passed in tensor.